File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;
class Solution {
    public int[] solution(int N, int[] stages) {
        PriorityQueue<double[]> pq = new PriorityQueue<double[]>((a,b) -> {
            if (a[1] == b[1]) {
                return Double.compare(a[0], b[0]);
            } else {
                return Double.compare(b[1], a[1]);
            }
        });
        Map<Integer, Integer> map = new HashMap<>();
        for (int i=0; i<stages.length; i++) {
            map.put(stages[i], map.getOrDefault(stages[i], 0) + 1);
        }

        for (int i=1; i<=N; i++) {
            int c = map.getOrDefault(i, 0);
            int psum = 0;
            for (int j=i; j<=N+1; j++) {
                psum += map.getOrDefault(j, 0);
            }
            pq.add(new double[] {(double)i, psum == 0 ? 0 : (double)c/psum});
        }
        
        int[] answer = new int[N];
        int len = pq.size();
        for (int i=0; i<len; i++) {
            answer[i] = (int)(pq.poll()[0]);
        }
        return answer;
    }
}

File: 프로그래머스/1/42889. 실패율/실패율.js
function solution(N, stages) {
    let answer = [];
    let counts = new Array(N+2).fill(0);
    let totals = new Array(N+2).fill(0);
    
    for (s of stages) {
        for (let i=1; i<=s; i++) {
            counts[i]++;
        }
        totals[s]++;
    }
    
    for (let i=1; i<=N; i++) {
        answer.push([i, totals[i] / counts[i]]);
    }
    
    answer.sort((a,b) => {
        if (a[1] === b[1]) {
            return a[0] - b[0];
        } else {
            return b[1] - a[1];
        }
    })
    
    return answer.map((x) => x[0]);
}